src/typesystem.hpp

```diff
@@ -67,4 +62,26 @@ struct TypeSystemCore {
 	MonoId inst_impl(MonoId mono, std::unordered_map<MonoId, MonoId> const& mapping);
 	MonoId inst_with(PolyId poly, std::vector<MonoId> const& vals);
 	MonoId inst_fresh(PolyId poly);
+
+	TypeFunctionData& type_function_data_of(MonoId);
+	TypeFunctionId new_tf_var();
+	void unify_tf(TypeFunctionId, TypeFunctionId);
+private:
+	int compute_new_argument_count(
+	    TypeFunctionData const& a_data, TypeFunctionData const& b_data) const;
+
+	void point_tf_at_another(TypeFunctionId a, TypeFunctionId b);
```

Comment:
Shouldn't this be called something like "unify_right"? I'm not very sure, I find the name to be too extravagant. I guess that _is_ the action that we want to perform at a high level so It's fine.

Reply:
Yeah, I had some back and forth with this one, but in the end failed to come up with anything better :confused:

Reply:
Ah that's okay, the name is explanatory.